### PR TITLE
Switch to using enum-dispatch to switch between UDP and Quic

### PR DESCRIPTION
**Pull request by ryleung-solana**
#### Problem
We currently use repetitive code to switch between UDP and Quic, when we could be using enum-dispatch

#### Summary of Changes
Switch to using enum dispatch

Fixes #

### CODE REVIEW

File: client/src/quic_client.rs [hunk 1]
```diff
@@ -66,6 +67,7 @@ struct QuicClient {
 
 pub struct QuicTpuConnection {
     client: Arc<QuicClient>,
+    connection_stats: Arc<ConnectionCacheStats>,
```
Comment (t-nelson) [inline client/src/quic_client.rs:70]: sticking a cache implementation detail into the connection doesn't feel great.  can you break out this refactor to it's own commit so we can consider it in isolation?

not having all of the plumbing changes to sift through will make review the rest easier too

Comment (ryleung-solana) [reply client/src/quic_client.rs]: This was done in another PR.

File: client/src/connection_cache.rs [hunk 2]
```diff
@@ -384,114 +382,20 @@ fn get_connection(addr: &SocketAddr) -> (Connection, Arc<ConnectionCacheStats>)
     connection_cache_stats
         .get_connection_ms
         .fetch_add(get_connection_measure.as_ms(), Ordering::Relaxed);
-    (connection, connection_cache_stats)
```
Comment (lijunwangs) [inline client/src/connection_cache.rs:387]: What is the point of updating the connection_cache_stats if we are not returning it?

**Review by lijunwangs — APPROVED** [review]
Looks good

File: client/src/quic_client.rs [hunk 3]
```diff
@@ -173,76 +171,65 @@ impl TpuConnection for QuicTpuConnection {
...
         Ok(())
     }
 
-    fn send_wire_transaction_async(
-        &self,
-        wire_transaction: Vec<u8>,
-        stats: Arc<ClientStats>,
-    ) -> TransportResult<()> {
+    fn send_wire_transaction_async(&self, wire_transaction: Vec<u8>) -> TransportResult<()> {
+        let stats = Arc::new(ClientStats::default());
```
Comment (t-nelson) [inline client/src/quic_client.rs:199]: nit: doesn't seem like these really need to be an `Arc`?  could just declare the temporary inside the async closure and pass a mut ref to the client methods?

not a change for this PR, but consider a followup

**Review by t-nelson — APPROVED** [review]
we like the red lines

**Comment by pgarg66** [comment]
@ryleung-solana should this be cherry-picked to v1.10?

**Comment by ryleung-solana** [comment]
> @ryleung-solana should this be cherry-picked to v1.10?

Yes, probably, in order to make merging future changes into v1.10 easier. Let me do that.

**Comment by ryleung-solana** [comment]
Yeah, it's probably a good idea to cherry-pick this into v1.10 so future
changes that come in afterward can also be backported more easily.
